### PR TITLE
Add noShadowRestrictedNames lint rule

### PR DESCRIPTION
**Pull request by mattsjones**
I had to pull back from using `getRootScope().isGlobal` as it was picking up too many false negatives. Mostly around the usage of browser globals in node code. I reduced the matching set of globals to just the built-in JS ones. 

However, `./scripts/dev-rome lint` is still failing due the usage of `constructor` bindings internally within rome. We can reduce the set further but would be maintaining a custom list within the rule at that point. Thoughts?

### CODE REVIEW

File: packages/@romejs/js-compiler/transforms/lint/noShadowRestrictedNames.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    const bindings = getBindingIdentifiers(node.head);
+
+    if (node.id !== undefined) {
+      bindings.push(node.id);
+    }
+
+    return bindings;
+  }
+
+  if (node.type === 'VariableDeclaration' || node.param) {
```
Comment (sebmck) [inline packages/@romejs/js-compiler/transforms/lint/noShadowRestrictedNames.ts]: Could you make the `node.param` more specific? Looks like `node` will only ever be `CatchClause` here, easier to just check for that.

File: packages/@romejs/js-compiler/transforms/lint/noShadowRestrictedNames.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import {getBindingIdentifiers} from '@romejs/js-ast-utils';
+import {markup} from '@romejs/string-markup';
+
+function extractBindingIdentifiers(
+  node:
+    | VariableDeclaration
+    | FunctionDeclaration
+    | FunctionExpression
+    | CatchClause,
+): BindingIdentifier[] {
```
Comment (sebmck) [inline packages/@romejs/js-compiler/transforms/lint/noShadowRestrictedNames.ts]:
```suggestion
): Array<BindingIdentifier> {
```

Comment (mattsjones) [reply packages/@romejs/js-compiler/transforms/lint/noShadowRestrictedNames.ts]: lol knew I'd pick the least preferred one. Maybe an opportunity for an autofix later.

**Comment by mattsjones** [comment]
@sebmck I addressed your feedback, but still unsure about what restricted name list to use. If feel the current list is quite accurate but there is 5 instances of `constructor` being used as a local binding within rome. I'm happy to refactor those out if you agree to.

**Comment by sebmck** [comment]
Restricted names list looks good. We can come back to it when addressing how in general to handle global names, as it's something `undeclaredVariables` also encounters. Feel free to rename the `constructor` variables. Even just to something like `_constructor`.\

I also thought of a simpler way to do this lint:

```
const {node, context, scope} = path;

if (scope.node === node) {
  for (const [name, binding] of scope.getOwnBindings()) {
    if (restrictedNames.has(name)) {
      context.addNodeDiagnostic(binding.node, {
        category: 'lint/noShadowRestrictedNames',
        message: markup`Shadowing of global property <emphasis>${name}</emphasis>`,
      });
    }
  }
}

return node;
```

This makes it a lot more reliable since it covers conditions that the lint does, and removes the need for the node checks and will automatically support anything we add to scope in the future.

File: packages/@romejs/js-compiler/transforms/lint/noShadowRestrictedNames.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+const restrictedNames = new Set([...builtin, ...es5, ...es2015, ...es2017]);
+
+export default {
+  name: 'noShadowRestrictedNames',
+  enter(path: Path): TransformExitResult {
+    const {node, context, scope} = path;
+
+    if (scope.node === node) {
+      for (const [name, binding] of scope.getOwnBindings()) {
```
Comment (mattsjones) [inline packages/@romejs/js-compiler/transforms/lint/noShadowRestrictedNames.ts]: `scope.getOwnBindings()` is returning an empty Map for the `CatchClause`. It seems to be setting the binding correctly within `packages/@romejs/js-compiler/scope/evaluators/CatchClause.ts` though.

Comment (mattsjones) [reply packages/@romejs/js-compiler/transforms/lint/noShadowRestrictedNames.ts]: Ok, so I dig some digging. Turns out the `CatchClause` evaluator was incorrectly setting its bindings to its parents scope. I've [fixed this](https://github.com/facebookexperimental/rome/pull/120/commits/d05d831330ded5b744a5e3e340a111fc476a1c88) but now `lint/unusedVariables` is triggering on all catch clauses as it doesn't handle them at all. Once I've added `CatchClause` support to `lint/unusedVariables` I'll post back here. 

Or if I'm completely wrong let me know 😉

Comment (sebmck) [reply packages/@romejs/js-compiler/transforms/lint/noShadowRestrictedNames.ts]: Aha, yeah that would be the case. Let's mark all catch clause bindings as always referenced in `unusedVariables`. Pretty sure we have support for the ES syntax that allows you to omit the `catch` parameter but I'm not sure how well supported it is.

Comment (mattsjones) [reply packages/@romejs/js-compiler/transforms/lint/noShadowRestrictedNames.ts]: I agree for single param. Probably should support deconstructors though. Will update soon. 👍

**Comment by mattsjones** [comment]
@sebmck This one is ready to go now 👍

**Comment by sebmck** [comment]
Looks good, thank you!

**Pull request merged**